FilterList model and endpoints

lemonsaurus opened the pull request:
This pull request introduces a new model, FilterList, and the prerequisite viewset, URL endpoint, migrations, serializer and tests.

The intention of this model is to store things that we will either whitelist or blacklist. For example, we'll be storing file formats, guild invite IDs, and other things that are currently in the bot config file, but that we want to be able to modify during runtime.

Eventually, we may set up automation where this table is synchronized against stuff like the [Awesome Discord Programming List](https://github.com/mhxion/awesome-discord-communities).

**Review style: commit-by-commit.**

Closes #305.

## Routes
### GET /bot/filter-lists
Returns all filterlist items in the database.

#### Response format
```js
[
    {
        'id': "2309268224",
        'created_at': "01-01-2020 ...",
        'updated_at': "01-01-2020 ...",
        'type': "file_format",
        'allowed': 'true',
        'content': ".jpeg",
        'comment': "Popular image format.",
    },
    ...
]
```

#### Status codes
- 200: returned on success
- 401: returned if unauthenticated

### GET /bot/filter-lists/\<id:int\>
Returns a specific FilterList item from the database.

#### Response format
```js
{
    'id': "2309268224",
    'created_at': "01-01-2020 ...",
    'updated_at': "01-01-2020 ...",
    'type': "file_format",
    'allowed': 'true',
    'content': ".jpeg",
    'comment': "Popular image format.",
}
```

#### Status codes
- 200: returned on success
- 404: returned if the id was not found.

### GET /bot/filter-lists/get-types
Returns a list of valid list types that can be used in POST requests.

#### Response format
```js
[
    ["GUILD_INVITE","Guild Invite"],
    ["FILE_FORMAT","File Format"],
    ["DOMAIN_NAME","Domain Name"],
    ["FILTER_TOKEN","Filter Token"]
]
```

#### Status codes
- 200: returned on success

### POST /bot/filter-lists
Adds a single FilterList item to the database.

#### Request body
```js
{
   'type': str,
   'allowed': bool,
   'content': str,
   'comment': Optional[str],
}
```

#### Status codes
- 201: returned on success
- 400: if one of the given fields is invalid

### DELETE /bot/filter-lists/\<id:int\>
Deletes the FilterList item with the given `id`.

#### Status codes
- 204: returned on success
- 404: if a tag with the given `id` does not exist

Commit:
Rename utils.py to mixins.py. More precise.

https://github.com/python-discord/site/issues/305

Commit:
Add a mixin for adding created and updated times.

https://github.com/python-discord/site/issues/305

Commit:
Add the AllowList model and serializer.

This is the model which we will use for items that are either
blacklisted or whitelisted.

https://github.com/python-discord/site/issues/305

Commit:
Merge branch 'master' into whitelist_system

Commit:
Allowlist viewset.

https://github.com/python-discord/site/issues/305

Commit:
Minor fixes for imports and __init__ files.

https://github.com/python-discord/site/issues/305

Commit:
Set up url forwarding for the viewset.

https://github.com/python-discord/site/issues/305

Commit:
Add a migration for the new AllowList model.

https://github.com/python-discord/site/issues/305

Commit:
Add a UniqueConstraint to prevent duplicates.

https://github.com/python-discord/site/issues/305

Commit:
Return id from the AllowListSerializer.

https://github.com/python-discord/site/issues/305

Commit:
Fix a bug in an old migration.

https://github.com/python-discord/site/issues/305

Commit:
Add tests for the AllowList model.

https://github.com/python-discord/site/issues/305

Commit:
Improve some docstrings.

https://github.com/python-discord/site/issues/305

Commit:
Fix some broken tests.

The test_utils_account.py tests were never running, because the folder
they were in had no __init__.py file.

The test_models.py file was failing because it had an outdated import of
the ModelReprMixin, which has moved to a new file.

https://github.com/python-discord/site/issues/305

Commit:
Rename AllowList to AllowDenyList.

https://github.com/python-discord/site/issues/305

Commit:
100% coverage for account.py

https://github.com/python-discord/site/issues/305

Commit:
100% branch coverage for account.py

https://github.com/python-discord/site/issues/305

lemonsaurus committed:
Merge branch 'master' into whitelist_system

Commit:
Add an endpoint for getting AllowDenyList types.

https://github.com/python-discord/site/issues/305

Commit:
Add another AllowDenyList field, 'comment'.

This will be used to describe or justify the entries in the blacklist or
whitelist, for example for the guild name in the case of guild invite
IDs, so that we have some context when we're displaying the list.

https://github.com/python-discord/site/issues/305

Commit:
Simplify AllowDenyListType names.

GUILD_INVITE_ID -> GUILD_INVITE
WORD_WATCHLIST -> FILTER_TOKEN

MarkKoz commented on Pipfile.lock:
```diff
@@ -666,23 +645,20 @@
                 "sha256:0d94e0e05a19e57a99444b6ddcf9a6eb2e5c68d3ca1e98e90707af8152c90a92",
                 "sha256:35b2d75ee967ea93b55750aa9edbbf72813e06a66ba54438df2cfac9e3c27fc8"
             ],
-            "markers": "python_version >= '2.7' and python_version not in '3.0, 3.1, 3.2, 3.3'",
```
What's with all of these being removed? Maybe it's something added by a more recent version of pipenv than what you used to lock?

lemonsaurus replied:
hmm. interesting. I'm not sure. I can try to update pipenv and relock?

lemonsaurus replied:
Fixed in 7f00c53

MarkKoz commented on pydis_site/apps/api/viewsets/bot/allow_deny_list.py:
```diff
@@ -0,0 +1,83 @@
+    queryset = AllowDenyList.objects.all()
+
+    @action(detail=False, methods=["get"])
+    def get_types(self, _: Request) -> Response:
```
This endpoint does not appear to be documented above.

lemonsaurus replied:
Resolved in 53e4dfe

MarkKoz commented on pydis_site/apps/api/tests/test_allowlists.py:
```diff
@@ -0,0 +1,112 @@
+
+        for case in cases:
+            response = self.client.post(URL, data=case)
+            self.assertEqual(response.status_code, 400)
```
This should use a subtest. I assume Django supports those.

lemonsaurus replied:
resolved in 99fc3cd

MarkKoz commented on pydis_site/apps/api/tests/test_allowlists.py:
```diff
@@ -0,0 +1,112 @@
+
+    def test_name_gets_deleted(self):
+        response = self.client.delete(f"{URL}/{self.png_format.id}")
+        self.assertEqual(response.status_code, 204)
```
These seem redundant.

lemonsaurus replied:
resolved in 99fc3cd

Commit:
Relock with a newer version of pipenv

MarkKoz commented on pydis_site/apps/api/serializers.py:
```diff
@@ -97,6 +97,16 @@ class Meta:
+        """Metadata defined for the Django REST Framework."""
+
+        model = AllowList
+        fields = ('created_at', 'updated_at', 'type', 'allowed', 'content')
```
I'm rusty with DRF. Will it allow `created_at` and `updated_at` to be set/changed even though they're auto fields? If yes, we should prevent that. Also, why are they listed here, but not the `id`?

lemonsaurus replied:
`id` is added in a later commit.

I've tried to POST requests with `updated_at` set to a datetime ten years in the past, but it seems to just ignore it and set it to what it should be anyway, so I don't think it's possible to set them. I think it's overridden by the auto field logics, or that the auto stuff happens later in the lifecycle.

SebastiaanZ replied:
I think that this has less to do with DRF and more to do with Django: According to the documentation, `auto_now` and `auto_now_add` fields just ignore any value that you try to pass to it. This is what `auto_now_add` says:

> Note that the current date is always used; it’s not just a default value that you can override. So even if you set a value for this field when creating the object, it will be ignored. 

The comment for `auto_add` is similar, but does not include the `creating the object` line: It always just sets the current datetime when updating the model and will ignore any custom value you try to pass to it.

Commit:
Document the get_types endpoint.

Commit:
Minor changes to tests, use subTest.

Commit:
Rename AllowDenyList to FilterList

Commit:
Handle unique validator in DRF, not Django.

I was handling this in a Django vanilla kind of way, which was causing
the constraint to return a 500 instead of a 400. This changes the
approach to use the DRF way, and makes it return 400.

It doesn't actually change the way anything behaves, other than
returning the right status code.

Commit:
Add a test for checking duplicates.

Commit:
Merge branch 'master' into whitelist_system

Commit:
Fix multiple leafs for migration graph.

Commit:
Add a constraint in the DB model as well.

This really should've been handled automatically by DRF, and in the
future, it will be. But for now, we need to have constraints both on the
serializer (to get status code 400), and on the model (to prevent direct
database constraint violations).

See https://github.com/encode/django-rest-framework/issues/7173

SebastiaanZ commented on pydis_site/apps/api/migrations/0025_allow_custom_inserted_at_infraction_field.py:
```diff
@@ -14,6 +14,6 @@ class Migration(migrations.Migration):
             model_name='infraction',
             name='inserted_at',
-            field=models.DateTimeField(default=datetime.datetime.utcnow, help_text='The date and time of the creation of this infraction.'),
+            field=models.DateTimeField(default=timezone.now(), help_text='The date and time of the creation of this infraction.'),
```
You should provide the callable so Django can use it to generate a default when it needs it. If you call it in the definition, it will be a single, static default value.

```suggestion
            field=models.DateTimeField(default=timezone.now, help_text='The date and time of the creation of this infraction.'),
```

lemonsaurus replied:
good catch.

SebastiaanZ reviewed (dismissed):
Alright, that's quite a lot of commits and files, but it looks solid. I've also ran some manual tests locally by polling the endpoints; looks good. One requested change to prevent us from using a wrong default value somewhere.

lemonsaurus committed:
Provide callable, not return value.

Co-authored-by: Sebastiaan Zeeff <33516116+SebastiaanZ@users.noreply.github.com>

SebastiaanZ reviewed (approved):
Nice work. Adding in the rework/fix/kaizen of the social account tests made it a slightly less focused PR, but I doubt fixing that would have otherwise been a major priority. Better to get it done.

Commit:
Add a migration for all the existing data.

This will populate the table with domain names, guild invites, filter
tokens and file formats.

Commit:
Delete FilterList objects for tests.

Now that we have a migration that adds data, we can no longer have tests
that operate on the assumption that the database is going to be empty.

So, we're now clearing that table before these tests run.
